jellyfin/jellyfin-androidtv
# Overhaul Preference typing / handling

**Changes**
This work came out of my trying to get my head around how Java/Kotlin generics interact. It started off as a small prototype but I saw benefits to the codebase so carried on.

Fundamentally this change boils down to:
- Use SealedClass for Preference Types instead of generics on the library side
```
    Use a list of sealed classes for preference types. This gives a number
    of advantages:
    
    - A strict list of accepted types is known at compile-time
    - The compiler can warn about missing clauses in when branches
    - We can elide throwing, as we know the types creatable
    - Interface is a lot cleaner, as we don't have nested generics / enum
    cases
    - Don't have to deal with the quirks of Java type erasure for all but Enum types
```
- Extract common functionality into a new abstract class `BasicPreferenceStore`:

```
    Unifies the common bits, this will make it significantly easier to write
    future stores.
    It also means the most risky parts (de-serializing enums and dynamic
    dispatch) are kept in one place. Making debugging / fixes significantly
    easier.
```

**Testing**
- Open preferences in the library selection window
- Change some around, close and open the app. Make sure they are saved.
- Do the same for preferences in something like a TV library (as Display Settings are their own model)
